pystac/extensions/ext.py

```diff
@@ -0,0 +1,290 @@
+
+    def add(self, name: EXTENSION_NAMES) -> None:
+        _get_class_by_name(name).add_to(self.stac_object)
+
+    def remove(self, name: EXTENSION_NAMES) -> None:
+        _get_class_by_name(name).remove_from(self.stac_object)
+
+    @property
+    def classification(self) -> ClassificationExtension[pystac.Item]:
+        return ClassificationExtension.ext(self.stac_object)
+
+    @property
+    def cube(self) -> DatacubeExtension[pystac.Item]:
+        return DatacubeExtension.ext(self.stac_object)
+
+    @property
+    def eo(self) -> EOExtension[pystac.Item]:
+        return EOExtension.ext(self.stac_object)
+
+    @property
+    def grid(self) -> GridExtension:
+        return GridExtension.ext(self.stac_object)
+
+    @property
+    def mgrs(self) -> MgrsExtension:
+        return MgrsExtension.ext(self.stac_object)
+
+    @property
+    def pc(self) -> PointcloudExtension[pystac.Item]:
+        return PointcloudExtension.ext(self.stac_object)
+
+    @property
+    def proj(self) -> ProjectionExtension[pystac.Item]:
+        return ProjectionExtension.ext(self.stac_object)
+
+    @property
+    def sar(self) -> SarExtension[pystac.Item]:
+        return SarExtension.ext(self.stac_object)
+
+    @property
+    def sat(self) -> SatExtension[pystac.Item]:
+        return SatExtension.ext(self.stac_object)
+
+    @property
+    def storage(self) -> StorageExtension[pystac.Item]:
+        return StorageExtension.ext(self.stac_object)
+
+    @property
+    def table(self) -> TableExtension[pystac.Item]:
+        return TableExtension.ext(self.stac_object)
+
+    @property
+    def timestamps(self) -> TimestampsExtension[pystac.Item]:
+        return TimestampsExtension.ext(self.stac_object)
+
+    @property
+    def version(self) -> VersionExtension[pystac.Item]:
+        return VersionExtension.ext(self.stac_object)
+
+    @property
+    def view(self) -> ViewExtension[pystac.Item]:
+        return ViewExtension.ext(self.stac_object)
+
+    @property
+    def xarray(self) -> XarrayAssetsExtension[pystac.Item]:
+        return XarrayAssetsExtension.ext(self.stac_object)
+
+
+class _AssetExt(Generic[T]):
+    stac_object: T
+
+    def has(self, name: EXTENSION_NAMES) -> bool:
+        if self.stac_object.owner is None:
+            raise pystac.STACError(
+                f"Attempted to add extension='{name}' for an Asset with no owner. "
+                "Use Asset.set_owner and then try to add the extension again."
+            )
+        else:
+            return cast(
+                bool, _get_class_by_name(name).has_extension(self.stac_object.owner)
+            )
+
+    def add(self, name: EXTENSION_NAMES) -> None:
+        if self.stac_object.owner is None:
+            raise pystac.STACError(
+                f"Attempted to add extension='{name}' for an Asset with no owner. "
+                "Use Asset.set_owner and then try to add the extension again."
+            )
```

Comment:
In the past, this was a real pain to work with, and ended up with strange workarounds. e.g in some stactools packages it is common to have a metadata mapping class, preparing assets for the stac producer to add to the item. It currently cannot enrich the assets with any metadata from extensions.
I would be very glad if it was somehow possible to add extensions to assets, even if they're not yet added to their owners yet.
A suggestion: maybe storing all extension URIs in a separate list, which gets added to the parents list once added there.

Reply:
This is a slightly different issue than what this PR addresses, and is tracked here: https://github.com/stac-utils/pystac/issues/793. I 100% agree with you for the record.